nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,227 @@
+}
+
+func getPrecheckErrors(m interface{}, typeParam string) ([]nsxModel.UpgradeCheckFailure, error) {
+	connector := getPolicyConnector(m)
+	client := pre_upgrade_checks.NewFailuresClient(connector)
+	resultList, err := client.List(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, &typeParam, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return resultList.Results, nil
+}
+
+func setFailedPrechecksInSchema(d *schema.ResourceData, precheckWarnings []nsxModel.UpgradeCheckFailure) error {
+	var failedPrechecksList []map[string]interface{}
+	for _, warning := range precheckWarnings {
+		id := *warning.Id
+		message := *(warning.Message.Message)
+		elem := make(map[string]interface{})
+		elem["id"] = id
+		elem["message"] = message
+		failedPrechecksList = append(failedPrechecksList, elem)
+	}
+	return d.Set("failed_prechecks", failedPrechecksList)
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	d.SetId(id)
+	// 1. Upload upgrade bundle and wait for upload to complete,
+	// UC will be automatically upgraded after this step
+	err := uploadUpgradeBundle(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	// 2. Accept user agreement
+	err = acceptUserAgreement(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	// 3. Execute pre-upgrade checks
+	err = executePreupgradeChecks(m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	// 4. Check if there is error in prechecks
+	precheckFailures, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_FAILURE)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	if len(precheckFailures) > 0 {
+		return handleCreateError("NsxtUpgradePrepare", id, fmt.Errorf("Encounter errors in precheck %s", *(precheckFailures[0].Message.Message)))
+	}
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func resourceNsxtUpgradePrepareRead(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	precheckWarnings, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_WARNING)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrepare", id, err)
+	}
+	err = setFailedPrechecksInSchema(d, precheckWarnings)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrepare", id, err)
+	}
+	return nil
+}
+
+func resourceNsxtUpgradePrepareUpdate(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func resourceNsxtUpgradePrepareDelete(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func waitForBundleUpload(m interface{}, bundleID string) error {
+	connector := getPolicyConnector(m)
+	client := bundles.NewUploadStatusClient(connector)
+	// c := m.(nsxtClients)
+	// Todo:
+	// 1. utilize Min/Max RetryInterval in provider config
+	interval := 20000
+	maxRetry := 50
+	for i := 0; i < 50; i++ {
```

Comment:
Lets consider using `stateConf.WaitForState()` tool from terraform sdk for consistency